js/classes/entities/foreground/dialog_bubble.js

```diff
@@ -34,6 +42,164 @@ class DialogBubble {
+        let textStart = new Vector(
+            (Camera.SIZE.x - DialogBubble.SCALED_SIZE.x) / 2 + (0.1 * DialogBubble.SCALED_SIZE.x),
+            Camera.SIZE.y - DialogBubble.SCALED_SIZE.y + (0.1 * DialogBubble.SCALED_SIZE.x) - (5 * DialogBubble.SCALE) + 17
+        );
+        // The jump in pixels after each line!
+        const textJump = new Vector(0, 40); // NOTE: this is based off a text size of 34!
+        CTX.fillStyle = "black";
+        CTX.font = FONT.VT323_NORMAL;
+        this.visibleLines.forEach((line) => {
+            CTX.fillText(line, textStart.x, textStart.y);
+            textStart = Vector.add(textStart, textJump);
+        });
+    };
+
+    /**
+     * As a dialog bubble may be added/removed several times from the GameEngine, this resets its fields as if it were reconstructed,
+     * making it draw properly and not remove itself from the world immediately.
+     */
+    refresh() {
+        this.visibleLines = [];
+        for (let i = 0; i < this.targetLines.length; i++) {
+            this.visibleLines.push("");
+        }
+        this.currentLine = 0;
+        this.removeFromWorld = false;
+    };
+
+    /**
+     * This is called when the user continues the conversation while text is still being typed.
+     * It does not progress the conversation, but does finish the typing.
+     */
+    finishTyping() {
+        for (let i = 0; i < this.targetLines.length; i++) {
+            this.visibleLines[i] = this.targetLines[i];
+        }
+        this.currentLine = this.targetLines.length;
+    }
+
+    /**
+     * This is going to split the text into multiple lines (up to 6!) which will fit on the dialog bubble.
+     * @param {string} text The entirety of the text which you want to split
+     * @returns An array of strings guaranteed to fit on one line on the text bubble.
+     * @throws An error if text is too long to fit on a single bubble (6 lines!).
+     */
+    static splitText(text) {
+        CTX.fillStyle = "black";
+        CTX.font = FONT.VT323_NORMAL;
+
+        const words = text.split(' ');
+        let wordIndex = 0;
+        const lines = [];
+        let currLine = "";
+
+        const maxLineWidth = DialogBubble.SCALED_SIZE.x * 0.8;
+
+        // While there are still words to account for:
+        while (wordIndex < words.length) {
+            let space = "";
+            // While this line can still fit more words.
+            while (CTX.measureText(currLine + space + words[wordIndex]).width <= maxLineWidth) {
+                currLine = currLine + space + words[wordIndex];
+                space = " ";
+                wordIndex++;
+                if (wordIndex === words.length) break;
+            }
+            // Now, the next word cannot fit on this line.
+            // Append currLine to lines.
+            lines.push(currLine);
+            currLine = "";
+        }
+        if (lines.length > 6) throw new Error("Cannot fit the following string into a single dialog bubble: " + text);
```

Comment:
Maybe make a MAX_LINES constant? I'll stop suggesting static constants now sorry lol.

Reply:
This one is definitely a good idea. Thanks!